cli/util/uploader.go

```diff
@@ -0,0 +1,360 @@
+			*object.NewAttribute("prevHash", block.PrevHash.Reverse().String()),    // Previous block hash in LE form
+			*object.NewAttribute("time", strconv.FormatUint(block.Timestamp, 10)),  // Timestamp (milliseconds precision)
+		}
+		err = uploadBlock(ctx.Context, *p, signer, acc.PrivateKey().GetScriptHash(), containerID, bw.Bytes(), attrs)
+		if err != nil {
+			return cli.Exit(fmt.Sprintf("failed to upload block %d: %v", blockIndex, err), 1)
+		}
+
+		if blockIndex%1000 == 0 {
+			fmt.Fprintln(ctx.App.Writer, "Successfully uploaded block:", blockIndex)
+		}
+	}
+	fmt.Fprintf(ctx.App.Writer, "Total number of blocks uploaded: %d\n", int(currentBlockHeight)-maxBlockIndex)
+	err = updateIndexFiles(ctx, *p, clientSDK, containerID, *acc, uint(currentBlockHeight), indexFileAttribute, attr)
+	if err != nil {
+		return cli.Exit(fmt.Sprintf("failed to update index files after upload: %v", err), 1)
+	}
+	fmt.Fprintln(ctx.App.Writer, "Upload completed successfully.")
+	return nil
+}
+
+// fetchMaxBlockIndex searches for the maximum block index in the container.
+func fetchMaxBlockIndex(ctx context.Context, clientSDK *client.Client, containerID cid.ID, priv *keys.PrivateKey, currentHeight uint, attributeKey string) (int, error) {
+	height := int(currentHeight)
+
+	searchBatch := func(start, end int) (int, error) {
+		prm := client.PrmObjectSearch{}
+		filters := object.NewSearchFilters()
+		filters.AddFilter(attributeKey, fmt.Sprintf("%d", start), object.MatchNumGE)
+		filters.AddFilter(attributeKey, fmt.Sprintf("%d", end), object.MatchNumLE)
+		prm.SetFilters(filters)
+
+		objectIDs, err := neofs.ObjectSearch(ctx, clientSDK, priv, containerID.String(), prm)
+		if err != nil {
+			return 0, fmt.Errorf("failed to search blocks from %d to %d: %w", start, end, err)
+		}
+		numOIDs := len(objectIDs)
+		if numOIDs == 0 {
+			return 0, nil // No blocks in this range, continue searching.
+		}
+
+		// Return the max block index found in this batch.
+		return start + numOIDs - 1, nil
+	}
+
+	for ; height > 0; height -= searchBatchSize {
+		startIndex := height - searchBatchSize + 1
+		if startIndex < 0 {
+			startIndex = 0
+		}
+
+		maxInBatch, err := searchBatch(startIndex, height)
+		if err != nil {
+			return 0, err
+		}
+
+		if maxInBatch > 0 {
+			return maxInBatch, nil
+		}
+	}
+	return -1, nil
+}
+
+// updateIndexFiles updates the index files in the container.
+func updateIndexFiles(ctx *cli.Context, p pool.Pool, clientSDK *client.Client, containerID cid.ID, account wallet.Account, currentHeight uint, attributeKey string, blockAttributeKey string) error {
+	indexFileSize := ctx.Uint("index-file-size")
+	fmt.Fprintln(ctx.App.Writer, "Updating index files...")
+	prm := client.PrmObjectSearch{}
+	filters := object.NewSearchFilters()
+	filters.AddFilter(attributeKey, fmt.Sprintf("%d", 0), object.MatchNumGE)
+	prm.SetFilters(filters)
+
+	objectIDs, err := neofs.ObjectSearch(ctx.Context, clientSDK, account.PrivateKey(), containerID.String(), prm)
```

Comment:
Make `neofs.ObjectSearch` work with both single client and client pool. It may be reached via converting argument to some `ObjectSearchInitter` interface containing a single method. Define this interface on the `neofs` package side.